rust/connlib/tunnel/src/peer.rs

```diff
@@ -4,14 +4,113 @@ use std::time::Instant;
 
 use bimap::BiMap;
 use chrono::{DateTime, Utc};
-use connlib_shared::messages::{ClientId, DnsServer, GatewayId, ResourceId};
+use connlib_shared::messages::{
+    gateway::Filter, gateway::Filters, ClientId, DnsServer, GatewayId, ResourceId,
+};
 use ip_network::IpNetwork;
 use ip_network_table::IpNetworkTable;
-use ip_packet::{MutableIpPacket, Packet};
+use ip_packet::ip::IpNextHeaderProtocols;
+use ip_packet::{IpPacket, MutableIpPacket, Packet};
+use rangemap::RangeInclusiveSet;
 
 use crate::client::IpProvider;
+use crate::utils::contains;
 
-type ExpiryingResource = (ResourceId, Option<DateTime<Utc>>);
+#[derive(Debug)]
+enum FilterEngine {
+    PermitAll,
+    PermitSome(FilterEngineInner),
+}
+
+#[derive(Debug)]
+struct FilterEngineInner {
+    udp: RangeInclusiveSet<u16>,
+    tcp: RangeInclusiveSet<u16>,
+    icmp: bool,
+}
+
+impl From<&Filters> for FilterEngine {
+    fn from(value: &Filters) -> Self {
+        FilterEngine::PermitSome(value.into())
+    }
+}
+
+impl FilterEngine {
+    fn new() -> FilterEngine {
+        Self::PermitSome(FilterEngineInner::new())
+    }
+
+    fn is_allowed(&self, packet: &IpPacket) -> bool {
+        match self {
+            FilterEngine::PermitAll => true,
+            FilterEngine::PermitSome(filter_engine) => filter_engine.is_allowed(packet),
+        }
+    }
+
+    fn permit_all(&mut self) {
+        *self = FilterEngine::PermitAll;
+    }
+
+    fn add_filters<'a>(&mut self, filters: impl Iterator<Item = &'a Filter>) {
+        match self {
+            FilterEngine::PermitAll => {}
```

Comment:
Should we debug_assert this? Or change to `PermitSome`? Silently dropping the filters seems like a source of bugs.